### PR TITLE
Add action hooks to WooCommerce email trigger function. Resolves #24641

### LINKED ISSUE #24641: REQUEST: New hook for order email trigger
Hi,

I'm trying to trigger an action after an email has been sent and need a hook for this. Ideally I'd like to hook into the "trigger" function which is found inside each email class. 

Tested with "order completed" email, but applicable to all order emails. 

**/woocommerce/includes/emails/class-wc-email-xxxx.php**

```
		/**
		 * Trigger the sending of this email.
		 *
		 * @param int            $order_id The order ID.
		 * @param WC_Order|false $order Order object.
		 */
		public function trigger( $order_id, $order = false ) {
			$this->setup_locale();

			if ( $order_id && ! is_a( $order, 'WC_Order' ) ) {
				$order = wc_get_order( $order_id );
			}

			if ( is_a( $order, 'WC_Order' ) ) {
				$this->object                         = $order;
				$this->recipient                      = $this->object->get_billing_email();
				$this->placeholders['{order_date}']   = wc_format_datetime( $this->object->get_date_created() );
				$this->placeholders['{order_number}'] = $this->object->get_order_number();
			}

			if ( $this->is_enabled() && $this->get_recipient() ) {
				$this->send( $this->get_recipient(), $this->get_subject(), $this->get_content(), $this->get_headers(), $this->get_attachments() );
			}

			$this->restore_locale();
		}
```

**SUGGESTED CHANGE:** 

If we hook into $this->send() then we can trigger an action based on TRUE / FALSE:

```
			if ( $this->is_enabled() && $this->get_recipient() ) {

				$result = (  $this->send( $this->get_recipient(), $this->get_subject(), $this->get_content(), $this->get_headers(), $this->get_attachments() ) ) ? 'success' : 'failure';
				
				do_action('woocommerce_order_email_trigger', $result, $order, $this);

			}
```

This would then allow us to add an action using the result (true / false), order & email object, for example inserting an order note to record which email was sent when, or triggering a follow up action.

### PR DESCRIPTION
**Pull request by EdithAllison**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds an action hook to the email trigger. This allows developers to run their own code after the sending of a WooCommerce email is triggered. Action hook includes send result (true / false) and the email object. 

Sample use case: Allows inserting a note into order when an email is sent so customer service can easily see which emails have been sent by the plugin to the customer. 

Closes #24641

### How to test the changes in this Pull Request:

1. Add action to the theme's functions.php file. A simple test case would be to write to debug log when email is being sent:

add_action('woocommerce_email_trigger_new_order_order', 'agentur_neworder', 20, 2 );

function agentur_neworder($result, $email) {
	error_log('NEW ORDER RESULT: ' . $result);
	error_log( print_r($email, true) );
}

2. Place a new order
3. Check debug log. If successful it will contain the result note and email object. 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> New action hook for email trigger function which allows developers to run their own code after Woo emails are triggered.

### CODE REVIEW

File: includes/emails/class-wc-email-cancelled-order.php [hunk 1]
```diff
@@ -90,7 +90,8 @@ public function trigger( $order_id, $order = false ) {
 			}
 
 			if ( $this->is_enabled() && $this->get_recipient() ) {
-				$this->send( $this->get_recipient(), $this->get_subject(), $this->get_content(), $this->get_headers(), $this->get_attachments() );
+				$result = ( $this->send( $this->get_recipient(), $this->get_subject(), $this->get_content(), $this->get_headers(), $this->get_attachments() ) ) ? true : false;
```
Comment (claudiosanches) [inline includes/emails/class-wc-email-cancelled-order.php:93]: By default `WC_Email->send()` returns boolean, so there's no need to any extra code:

```suggestion
				$result = $this->send( $this->get_recipient(), $this->get_subject(), $this->get_content(), $this->get_headers(), $this->get_attachments() );
```

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Hello @EdithAllison,

Thanks for your help, in generally looks good, nice choice for naming all those actions, but there's some small tweaks in the code missing before we can merge this PR.

**Comment by EdithAllison** [comment]
@claudiosanches Many thanks for considering this PR, completely agree with code simplification. I've updated the PR as suggested 👍

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Hello @EdithAllison,

I'm checking here, why not changing `WC_Email->send()` method instead? So will be required just one hook for all emails, since all email classes extend `WC_Email` as well.

**Comment by claudiosanches** [comment]
I wrote a new PR with just one hook that may help: https://github.com/woocommerce/woocommerce/pull/30123
I'm going to close this PR for now.
Thanks.

**Comment by EdithAllison** [comment]
@claudiosanches Noted. I didn't realised Woo was using its own email class, I wrongly assumed it was using WP core email functionality. Agree that a single change to Woo email class is a better implementation, thank you for taking the time to review and create a PR.

**Pull request closed without merging**